Add ops doc on enabling in-proxy metrics generation for 1.3 (#4957)
istio/istio.io: content/en/docs/ops/telemetry/in-proxy-service-telemetry/index.md

extensibility architecture for Istio, which will put Mixer-like functionality directly in the sidecar proxies.

The in-proxy generation of service-level metrics is a drop-in replacement for the HTTP metrics currently
generated by Mixer, namely: `istio_requests_total`, `istio_request_duration_*` and `istio_request_size`.

## Enable Service-level Metrics Generation in Envoy

To enable the generation of service-level metrics in Envoy, execute the following steps:

1. Disable calls to `istio-telemetry` in the mesh

    In order to prevent duplicate telemetry generation, turn off the use of the `istio-telemetry` service
    for telemetry generation.

    {{< text bash >}}
    $ helm template install/kubernetes/helm/istio --name istio --namespace istio-system --set mixer.telemetry.enabled=false --set mixer.policy.enabled=false
    {{< /text >}}

    Alternatively, comment out `mixerCheckServer` and `mixerReportServer` from the mesh config
    (the `istio` config map).

1. Enable the metadata exchange filter

    For the generation of service-level metrics, proxies must exchange workload metadata. This exchange
    is handled by a custom filter.

    {{< text bash >}}

>     $ kubectl -n istio-system apply -f https://github.com/istio/proxy/blob/master/extensions/stats/testdata/istio/metadata-exchange_filter.yaml

Reviewer: this needs the raw mode, not a link to the github UI

    {{< /text >}}

1. Enable the custom stats filter

    A custom filter must also be applied to actually generate the service-level metrics.

    {{< text bash >}}
    $ kubectl -n istio-system apply -f https://raw.githubusercontent.com/istio/proxy/master/extensions/stats/testdata/istio/stats_filter.yaml
    {{< /text >}}

1. See the results

    Visit Grafana "Istio Mesh" dashboard. It will display the same telemetry as before -- without
    any requests flowing through Istio's Mixer.

## Differences with Mixer-based generation

There are some small differences between the in-proxy generation of the service-level metrics and
the Mixer-based generation in the 1.3.0 release. Full feature-parity is a requirement before this
functionality can be considered stable.

In the meantime, please be aware of these differences:

> 1. The latency metric uses a new name and is recorded in different units. The new metric is called

Reviewer 2: ```suggestion
- The `istio_request_duration_seconds` latency metric has the new name: `istio_request_duration_milliseconds`. 
  The new metric uses milliseconds instead of seconds. We updated the Grafana dashboards to
  account for these changes.
```

>    `istio_request_duration_milliseconds` (instead of `istio_request_duration_seconds`). The Grafana

Reviewer 2: ```suggestion
```

>    dashboards are updated to account for the new metric.

Reviewer 2: ```suggestion
```

> 1. The `istio_request_duration_milliseconds` metric uses more granular buckets inside the proxy.

Reviewer 2: ```suggestion
- The `istio_request_duration_milliseconds` metric uses more granular buckets inside the proxy.
```

>    This results in lower latency measurements in the histograms than before.

Reviewer 2: ```suggestion
 This change leads to lower latency measurements in the histograms than previously observed.
```

> ## Performance impact

Reviewer 2: Please be extra careful when talking about performance. I've left suggestions to be as specific and clear as possible.
@askmeegs @oaktowner PTAL and ensure this content won't lead to any confusion.

Author: roger.

{{< warning >}}

> The focus of the initial release of this functionality has not been on performance optimization.

Reviewer 2: ```suggestion
This release doesn't focus on performance optimization for this functionality.
```

> We have several planned performance-related improvements in the pipeline at this time.

Reviewer 2: ```suggestion
We are working on several performance-related improvements for this experimental feature.
```

> This feature will not be considered for promotion to Beta or Stable status until our performance

Reviewer 2: ```suggestion
We won't consider this feature for promotion to **Beta** or **Stable** [status](/about/feature-stages/#feature-phase-definitions)
until we realize our performance goals.
```

> goals have been realized.

Reviewer 2: ```suggestion
```

> 

Reviewer 2: ```suggestion

The performance of your mesh depends on your configuration. To learn more, see our [performance best practices post](/blog/2019/performance-best-practices/).

```

{{< /warning >}}

> 

Reviewer 2: ```suggestion

If you are encountering performance issues with this feature, consider the following actions to improve it:

```

Reviewer: I don't think this makes sense. The list below is what the result is of turning this on, not actions to take

Author: @howardjohn agree.

> 1. The `istio-telemetry` deployment can be switched off. This should save ~0.5 vCPU per 1000 rps

Reviewer 2: ```suggestion
- You can disable the `istio-telemetry` deployment. Disabling it can potentially save ~0.5 vCPU per 1000 rps
```

>    of mesh traffic. This also halves the CPU consumed by Istio while collecting Istio standard

Reviewer 2: ```suggestion
   of mesh traffic. This change can potentially halve the CPU resources Istio consumes while collecting the default
```

   metrics.

> 1. The new filters together use 10% less CPU for `istio-proxy` containers than the Mixer filter.

Reviewer 2: ```suggestion
- All new filters together use 10% less CPU resources for the `istio-proxy` containers than the Mixer filter.
```

> 1. The new filters add ~5ms P90 latency (1000 rps) as compared to a generic Envoy in the current

Reviewer: Can we compare this to the current Mixer usage? This makes it sound like it is 5ms worse than Mixer, not 5ms worse than no metrics

Author: Reworded. PTAL.

> 1. The new filters add ~5ms P90 latency (1000 rps) as compared to a generic Envoy in the current

Reviewer 2: ```suggestion
- When compared to the current implementation, the new filters only add ~5ms P90 latency per
  1000 rps compared to the performance we could measure when we configured the mesh to
  collect no metrics.
```

>    implementation.

Reviewer 2: ```suggestion
```

> ## Known limitations in 1.3

Reviewer 2: ```suggestion
## Currently known limitations
```

> 1. Only Prometheus export of these metrics is supported.

Reviewer 2: ```suggestion
- You can only export these metrics to Prometheus.
```

> 1. No support for TCP metrics generation is provided.

Reviewer 2: ```suggestion
- We provide no support to generate TCP metrics.
```

> 1. No proxy-side customization or configuration of the generated metrics is provided.

Reviewer 2: ```suggestion
- We provide no proxy-side customization or configuration of the generated metrics.
```

> ## Implementation Details

Reviewer 2: I don't think we need to have this content here. Implementation details should not be part of operation guides.
I'm leaving suggestion on how to fix the text but I advice removing this section altogether.
```suggestion
## Implementation details
```

> The 1.3.0 implementation uses the WASM sandbox API, but does not run inside a WASM VM. It is

Reviewer 2: ```suggestion
The implementation in Istio 1.3 uses the `WASM` sandbox API, but it doesn't run inside a `WASM VM`.
```

> natively-compiled in Envoy using `NullVM`. In subsequent releases, we will enable running

Reviewer 2: ```suggestion
The implementation is natively-compiled in Envoy using `NullVM`. We are working to enable running
```

> filters in the V8 WASM VM.

Reviewer 2: ```suggestion
filters in the `V8 WASM VM` in future releases.
```

> In 1.3, the filters are configured using the Istio Envoy Filter API. For the 1.4 release,

Reviewer 2: ```suggestion
Currently you configure the filters using the Istio Envoy filter API. One of our goals for Istio 1.4 is to
```

> we will introduce new ways of configuring extensions.

Reviewer 2: ```suggestion
introduce new ways to configure extensions.
```

---
title: Generate Istio Metrics Without Mixer
description: In-proxy generation of service-level metrics.
weight: 70
---

{{< warning >}}

This document describes **Alpha** quality features. They have been exposed *only* to enable further testing
and evaluation across the community.

The functionality described here is NOT considered stable. No performance or stability guarantees are provided
at this time.

{{< /warning>}}

Istio 1.3 adds experimental support to generate service-level HTTP metrics
directly in the Envoy proxies. This feature lets you continue to monitor your service meshes using the tools Istio provides

> without needing Mixer. This support is the first step towards the new

Reviewer 3: is there any stable link for the Mixer v2 plan / a high level doc?

Author: Not published to the website. I also don't believe we have a full up-to-date doc that we can point to here. Stay tuned, however.

Author: I've removed the references here. That detail isn't needed.

extensibility architecture for Istio. This architecture puts Mixer-like functionality directly in the Envoy proxies.

The in-proxy generation of service-level metrics replaces the following HTTP metrics that Mixer currently

> generates:

Reviewer 3: worth adding that in-proxy does not currently support other metrics like latency?

- `istio_requests_total`
- `istio_request_duration_*`
- `istio_request_size`

## Enable Service-level Metrics Generation in Envoy

To generate service-level metrics directly in the Envoy proxies, follow these steps:

1.  To prevent duplicate telemetry generation, turn off the `istio-telemetry` service to disable calls to `istio-telemetry` in the mesh:

    {{< text bash >}}
    $ helm template install/kubernetes/helm/istio --name istio --namespace istio-system --set mixer.telemetry.enabled=false --set mixer.policy.enabled=false
    {{< /text >}}

        {{< tip >}}

>         Alternatively, you can comment out `mixerCheckServer` and `mixerReportServer` in your mesh's configuration file.

Reviewer 3: not clear what you mean by "mesh configuration file" - provide an example or a link?

Author: There was a mention previously of the `istio` configmap. I guess that got removed in one of the suggestions/edits. I'll see if I can find a good reference link.

Author: Updated with a link to the reference doc.

        {{< /tip >}}

1. To generate service-level metrics, the proxies must exchange {{< gloss >}}workload{{< /gloss >}} metadata. 
   A custom filter handles this exchange. Enable the metadata exchange filter with the following command:

    {{< text bash >}}
    $ kubectl -n istio-system apply -f https://raw.githubusercontent.com/istio/proxy/master/extensions/stats/testdata/istio/metadata-exchange_filter.yaml
    {{< /text >}}

> 1. To actually generate the service-level metrics, you must apply the custom stats filter:

Reviewer 3: why is there a semicolon after filter? should there be a command here?

>     A custom filter must also be applied to actually generate the service-level metrics.

Reviewer 3: this seems like a duplicate of the previous sentence

Author: removed.

    {{< text bash >}}
    $ kubectl -n istio-system apply -f https://raw.githubusercontent.com/istio/proxy/master/extensions/stats/testdata/istio/stats_filter.yaml
    {{< /text >}}

> 1. Go to the **Istio Mesh** dashboard Grafana. Verify that the dashboard displays the same telemetry as before but without

Reviewer 3: swap the words 'dashboard' and 'Grafana' ?

Author: done.

    any requests flowing through Istio's Mixer.

## Differences with Mixer-based generation

Small differences between the in-proxy generation and Mixer-based generation of service-level metrics
persist in Istio 1.3. We won't consider the functionality stable until in-proxy generation has full feature-parity with
Mixer-based generation.

Until then, please consider these differences:

- The `istio_request_duration_seconds` latency metric has the new name: `istio_request_duration_milliseconds`. 
  The new metric uses milliseconds instead of seconds. We updated the Grafana dashboards to
  account for these changes.
- The `istio_request_duration_milliseconds` metric uses more granular buckets inside the proxy.

>  This change leads to lower latency measurements in the histograms than previously observed.

Reviewer 3: I'm not clear on why more histogram buckets results in different latency observations. Please clarify

Author: reworded and avoided "lower" comparison.

## Performance impact

{{< warning >}}

This release doesn't focus on performance optimization for this functionality.

> We are working on several performance-related improvements for this experimental feature.

Reviewer 3: I'd remove this unless you have a link to a performance roadmap, otherwise seems too vague.

Author: Rephrased the entire section. PTAL.

{{< /warning >}}

1. The `istio-telemetry` deployment can be switched off. This should save ~0.5 vCPU per 1000 rps
   of mesh traffic. This also halves the CPU consumed by Istio while collecting Istio standard
   metrics.
- All new filters together use 10% less CPU resources for the `istio-proxy` containers than the Mixer filter.
1. The new filters add ~5ms P90 latency (1000 rps) as compared to a generic Envoy in the current
   implementation.

## Known limitations in 1.3

- You can only export these metrics to Prometheus.
- We provide no support to generate TCP metrics.
- We provide no proxy-side customization or configuration of the generated metrics.

## Implementation details

The implementation in Istio 1.3 uses the `WASM` sandbox API, but it doesn't run inside a `WASM VM`.
The implementation is natively-compiled in Envoy using `NullVM`. We are working to enable running
filters in the `V8 WASM VM` in future releases.

Currently you configure the filters using the Istio Envoy filter API. One of our goals for Istio 1.4 is to

> we will introduce new ways of configuring extensions.

Reviewer 3: I would rewrite this section to only include the implementation details for 1.3, and leave out 1.4 roadmap items.

Author: Just removed this section altogether. Not useful for the purpose of this doc.

## Differences

There are some small differences in the generation of the service-level metrics at the moment. Full feature-parity is a requirement before this
functionality can be considered stable. In the meantime, please be aware of these differences:

1. The `istio_request_duration_*` metric uses more granular buckets inside the proxy. This results in lower latency measurements in the histograms than before.
1. The new metric is called `istio_request_duration_milliseconds` (instead of `istio_request_duration_seconds`). The Grafana dashboard is updated to account for the new metric.

## Performance impact

{{< warning >}}

The focus of the initial release of this functionality has not been on performance optimization. We have several planned performance-related improvements
in the pipeline at this time.

This feature will not be considered for promotion to Beta or Stable status until our performance goals have been realized.

{{< /warning >}}

1. The `istio-telemetry` deployment can be switched off. This should save ~0.5 vCPU per 1000 rps of mesh traffic. This also halves the CPU consumed by Istio while collecting Istio standard metrics.

> 1. The new filters use 10% less CPU for `istio-proxy` containers than the Mixer filter.

Reviewer 4: The new filters `together`    , otherwise it is unclear if it is 10% less each or combined.

Author: done.

---
title: Generate Istio Metrics Without Mixer [Experimental]
description: How to enable in-proxy generation of HTTP service-level metrics.
weight: 70
---

{{< boilerplate experimental-feature-warning >}}

Istio 1.3 adds experimental support to generate service-level HTTP metrics
directly in the Envoy proxies. This feature lets you continue to monitor your
service meshes using the tools Istio provides without needing Mixer.

The in-proxy generation of service-level metrics replaces the following HTTP
metrics that Mixer currently generates:

- `istio_requests_total`
- `istio_request_duration_seconds`
- `istio_request_size`

> ## Enable Service-level Metrics Generation in Envoy

Reviewer 2: ```suggestion
## Enable service-level metrics generation in Envoy
```

  account for these changes.
- The `istio_request_duration_milliseconds` metric uses more granular buckets inside the proxy, providing
  increased accuracy in latency reporting.

## Performance impact

{{< warning >}}

As this work is currently experimental, our primary focus has been on establishing
the base functionality. We have identified several performance optimizations based
on our initial experimentation, and expect to continue to improve the performance
and scalability of this feature as it develops.

We won't consider this feature for promotion to **Beta** or **Stable** [status](/about/feature-stages/#feature-phase-definitions)
until performance and scalability assessments and improvements have been made.

The performance of your mesh depends on your configuration.
To learn more, see our [performance best practices post](/blog/2019/performance-best-practices/).

{{< /warning >}}

Here's what we've measured so far:

- All new filters together use 10% less CPU resources for the `istio-proxy` containers
  than the Mixer filter.
- The new filters add ~5ms P90 latency at 1000 rps compared to Envoy proxies
  configured with no telemetry filters.

> - If you are using `istio-telemetry` solely to generate service-level metrics, the

Reviewer 2: ```suggestion
- If you only use the `istio-telemetry` service to generate service-level metrics,
```

>   `istio-telemetry` deployment can be switched off. This should save ~0.5 vCPU per

Reviewer 2: ```suggestion
  you can switch off the `istio-telemetry` service. This could save up to ~0.5 vCPU per
```

>   1000 rps of mesh traffic. This also halves the CPU consumed by Istio while collecting Istio

Reviewer 2: ```suggestion
  1000 rps of mesh traffic, and could halve the CPU consumed by Istio while collecting
```
